src/plugins/intel_cpu/src/nodes/llm_mlp.cpp

```diff
@@ -0,0 +1,652 @@
+                            // the prefetch distance is increased to ensure by the time store happens
+                            // prefetch has done and no HW prefetcher is triggered
+                            auto* prefetch_dst = (m + 2 < M) ? (dst + 2 * strideD) : (dst);
+                            jit_reduce2bh_2(src0, src1, dst, prefetch_dst, work.BN);
+                        }
+                    }
+                } else {
+                    auto* src = workC.ptr<float>();
+                    auto* dst = dstC + work.n0;
+                    auto strideS = workC.stride(0);
+                    auto strideD = strideC / sizeof(*dst);
+                    for (int m = 0; m < M; m++, src += strideS, dst += strideD) {
+                        // the prefetch distance is increased to ensure by the time store happens
+                        // prefetch has done and no HW prefetcher is triggered
+                        auto* prefetch_dst = (m + 2 < M) ? (dst + 2 * strideD) : (dst);
+                        jit_reduce2bh_1(src, dst, prefetch_dst, work.BN);
+                    }
+                }
+            }
+        });
+    }
+
+    // gate & up are interleaved: 16 gates + 16 up
+    void runGateUp(uint8_t* pA, int strideA, int M, ov::bfloat16* dstC, int strideC) {
+        static GateUpCombine jit_gateup;
+        ov::parallel_nt_static(0, [&](const size_t ithr, const size_t nthr) {
+            auto& work = works[ithr];
+            auto& workC = getC(ithr);
+            if (work.BN > 0) {
+                work.run(M, pA, strideA, workC);
+                // K reduce is done, results of [M, BN] sub-block is ready in L2.
+                // combine Gate & Up
+                auto* src = workC.ptr<float>();
+                auto strideS = workC.stride(0);
+                auto* dst = dstC + (work.n0 / 2);  // important output is only half of the total N
+                auto strideD = strideC / sizeof(*dst);
+                for (int m = 0; m < M; m++, src += strideS, dst += strideD) {
+                    auto* prefetch_dst = (m + 1 < M) ? (dst + strideD) : (dst);
+                    jit_gateup(src, dst, prefetch_dst, work.BN);
+                }
+            }
+        });
+    }
+};
+
+struct QKVProj : public LLMMLP::Executor {
+    QKVProj() {}
+    std::vector<Work> works;
+
+    // q k v each have 1/3 or worker-thread
+    void setup(ov::bfloat16* wq, ov::bfloat16* wk, ov::bfloat16* wv, int N, int K) {
+        const int blk_K_size = 256;
+        // prepare weights, split N among threads
+        // in unit of 32
+        OPENVINO_ASSERT((N % 32) == 0);
+        OPENVINO_ASSERT((K % blk_K_size) == 0);
+        auto nthr = parallel_get_max_threads();
+        auto num_blk_N = N / 32;
+        auto num_blk_K = K / blk_K_size;
+        works.resize(nthr);
+
+        int stride = K * sizeof(*wq);
+
+        // every thread should do same amount of work, and some cores can be idle
+        auto valid_nthr = nthr / 3;
```

Comment:
Does this mean there will be at most 2 threads which have works?

Reply:
Good insight, this logic would fail when (nthr < 3).

Reply:
minimal nthr is checked in transformation to avoid failure here.